lib/environment/background/redirection.js

```diff
@@ -0,0 +1,86 @@
+/* @flow */
+
+import { regexes } from '../../utils/location';
+import { addListener } from './messaging';
+
+const state = {
+	enabled: false,
+	options: {},
+};
+let webReqListenerAdded: boolean = false;
+
+// this listener handles changes to the redirect module's state
+addListener('redirection', ({ type, data }) => {
+	if (type === 'updateState') {
+		state.enabled = data.enabled;
+		state.options = { ...data.options };
+
+		if (!webReqListenerAdded && state.enabled) {
+			chrome.webRequest.onBeforeRequest.addListener(
+				webReqListener,
+				{ urls: ['*://*.reddit.com/*'], types: ['main_frame'] },
+				['blocking']
+			);
+			webReqListenerAdded = true;
+		}
+
+		return true;
+	} else {
+		throw new Error(`Invalid redirection operation: ${type}`);
+	}
+});
+
+// this listener actually does the redirects
+function webReqListener({ url }) {
+	const { enabled, options } = state;
+	if (!enabled) return {};
+
+	const { origin: urlOrigin, pathname: urlPath } = new URL(url);
+
+	let name: string | void, currentSection: string | void;
+
+	// user profile landing page check
+	[, name, currentSection] = regexes.profile.exec(urlPath) || [];
```

Comment:
I wonder if this could be cleaner if the redirect interface was declarative. E.g. could a module add a redirect with a substitution: 

```
redirect.addRegex({
  pathnameRegex: /^\/user\/([\w\-]+)(?:\/)?\/?$/i,
  substitution: `/user/$1/${module.options.customFromProfileLandingPage || module.options.fromProfileLandingPage}`
})
```

Reply:
The latest commits include this change. Any module can now add/delete redirect rules like:

`
Redirection.updateRedirectBackgroundState([
    {from: (/^\/user\/([\w\-]+)(?:\/)?\/?$/i).toString(), fromType: 'abs', to: '/user/$1/hot'},
    {from: (/^\/r\/([\w\-]+)(?:\/)?\/?$/i).toString(), fromType: 'abs', to: null},
])
`

Reply:
Rough edges in the current implementation:

1. The home-page redirect rule (used to send users to old.reddit.com or np.reddit.com etc. instead of www.reddit.com) doesn't work when the user clicks on the reddit logo when on a reddit subdomain other than www

2. There should be some protection to prevent users getting into an infinite redirection loop by setting the redirection options a certain way. 

For example, setting fromHomePage to 'custom' and customFromProfileLandingPage to the empty string basically breaks reddit for the user

Reply:
3. Possible redirect-rule-name and redirect-rule-from collisions when other modules are adding redirect rules. Need some kind of warning system.